core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/VariableDescriptorAwareScoreDirector.java

```diff
@@ -0,0 +1,193 @@
+     * <li>Any mix of the above.</li>
+     * </ul>
+     * For the sake of variable listeners' efficiency, the change notification requires an index range that contains elements
+     * affected by the change. The range starts at {@code fromIndex} (inclusive) and ends at {@code toIndex} (exclusive).
+     * <p>
+     * The range has to comply with the following contract:
+     * <ol>
+     * <li>{@code fromIndex} must be greater than or equal to 0; {@code toIndex} must be less than or equal to the list variable
+     * size.</li>
+     * <li>{@code toIndex} must be greater than or equal to {@code fromIndex}.</li>
+     * <li>The range must contain all elements that are going to be changed.</li>
+     * <li>The range is allowed to contain elements that are not going to be changed.</li>
+     * <li>The range may be empty ({@code fromIndex} equals {@code toIndex}) if none of the existing list variable elements
+     * are going to be changed.</li>
+     * </ol>
+     * <p>
+     * {@link #beforeListVariableElementUnassigned(ListVariableDescriptor, Object)} must be called for each element
+     * that will be unassigned (removed from a list variable of one entity without being added
+     * to a list variable of another entity).
+     *
+     * @param variableDescriptor descriptor of the list variable being changed
+     * @param entity the entity owning the list variable being changed
+     * @param fromIndex low endpoint (inclusive) of the changed range
+     * @param toIndex high endpoint (exclusive) of the changed range
+     */
+    void beforeListVariableChanged(ListVariableDescriptor<Solution_> variableDescriptor, Object entity, int fromIndex,
+            int toIndex);
+
+    /**
+     * Notify the score director after a list variable changes.
+     * <p>
+     * The list variable change includes:
+     * <ul>
+     * <li>Changing position (index) of one or more elements.</li>
+     * <li>Removing one or more elements from the list variable.</li>
+     * <li>Adding one or more elements to the list variable.</li>
+     * <li>Any mix of the above.</li>
+     * </ul>
+     * For the sake of variable listeners' efficiency, the change notification requires an index range that contains elements
+     * affected by the change. The range starts at {@code fromIndex} (inclusive) and ends at {@code toIndex} (exclusive).
+     * <p>
+     * The range has to comply with the following contract:
+     * <ol>
+     * <li>{@code fromIndex} must be greater than or equal to 0; {@code toIndex} must be less than or equal to the list variable
+     * size.</li>
+     * <li>{@code toIndex} must be greater than or equal to {@code fromIndex}.</li>
+     * <li>The range must contain all elements that have changed.</li>
+     * <li>The range is allowed to contain elements that have not changed.</li>
+     * <li>The range may be empty ({@code fromIndex} equals {@code toIndex}) if none of the existing list variable elements
+     * have changed.</li>
+     * </ol>
```

Comment:
For `AbstractEasyMove`, there are additional restrictions on `beforeListVariableChange` and `afterListVariableChanged`. In particular:

- The `fromIndex` of `afterListVariableChanged` must match the `fromIndex` of its  `beforeListVariableChanged` call. Otherwise this will happen in the undo move
```
// beforeListVariableChanged(0, 3);
[1, 2, 3, 4]
change
[1, 2, 3]
// afterListVariableChanged(2, 3)
// Start Undo
// Undo afterListVariableChanged(2, 3)
[1, 2, 3] -> [1, 2]
// Undo beforeListVariableChanged(0, 3);
 [1, 2, 3, 4, 1, 2]
``` 
`VariableChangeRecordingScoreDIrector` should be able to detect this. In particular, instead of creating the ListVariableAfterChangeAction` using the `fromIndex` from the `afterListVariableChanged` call, it will use the `fromIndex` from the `beforeListVariableChanged` call (by tracking it in an identity hash map using the entity as key).

Reply:
This move is quickly becoming expensive. Now we add map overhead to the existing list overhead. I'll have to benchmark this and see how big of a deal it really is. Hopefully not much.